Sources/SwiftSyntaxBuilder/Syntax+StringInterpolation.swift

```diff
@@ -192,3 +228,215 @@ extension SyntaxExpressibleByStringInterpolation {
     self.init(stringLiteral: source)
   }
 }
+
+// MARK: ExpressibleByLiteralSyntax conformances
+
+extension Collection {
+  fileprivate func allIndices(where predicate: @escaping (Element) -> Bool) -> UnfoldSequence<Index, Index> {
+    sequence(state: startIndex) { i in
+      guard let newI = self[i...].firstIndex(where: predicate) else {
+        return nil
+      }
+      i = index(after: newI)
+      return newI
+    }
+  }
+}
+
+extension Substring: ExpressibleByLiteralSyntax {
+  public func makeLiteralSyntax() -> ExprSyntaxProtocol {
```

Comment:
I think of `makeLiteralSyntax()` as purely an implementation detail of `ExpressibleByLiteralSyntax`, not something users would call directly. If we want to expose this functionality for use outside of interpolations (probably a good idea!), I'd put an initializer on `ExprSyntax` or something and tell people to call that.

Reply:
Also, looking at the `StringLiteralExpr` initializer in particular, it has some advanced capabilities that this doesn't; for instance, you can specify a raw delimiter, and if you do then escape sequences in the `content` are not escaped. I assume we want to hang on to those capabilities.

Reply:
I just put this comment on `Substring.makeLiteralSyntax` because it's where I saw it doing the much of the implementation of the `StringLiteralExpr` convenience initializer. I too would consider  `makeLiteralSyntax` an implementation detail.

> for instance, you can specify a raw delimiter, and if you do then escape sequences in the content are not escaped

That's true, but then you could also just use the regular constructor if you *really* want that (providing a `StringLiteralSegments`), so I'm still not convinced it adds much over `ExpressibleByLiteralSyntax`. Its main use case was escaping the literal, IMO.

Anyway, I don't mind having both, but I would strongly prefer them to use the same escaping implementation (totally fine if you think yours handles more cases/is better).

Reply:
I'm working on bringing the existing implementation up to snuff so that the new methods can use that.